### PR TITLE
[NOD-1123] Implement banning

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: connmanager/connmanager.go [hunk 1]
```diff
@@ -112,3 +112,8 @@ func (c *ConnectionManager) connectionsLoop() {
 func (c *ConnectionManager) ConnectionCount() int {
 	return c.netAdapter.ConnectionCount()
 }
+
+// Ban prevents the given netConnection from connecting again
+func (c *ConnectionManager) Ban(netConnection *netadapter.NetConnection) {
+	c.netAdapter.Ban(netConnection)
```
Comment (someone235) [inline connmanager/connmanager.go:118]: Add a ticket to add it to persistence and apply it for only 24 hours

File: protocol/protocol.go [hunk 2]
```diff
@@ -57,6 +56,11 @@ func (m *Manager) startFlows(netConnection *netadapter.NetConnection, router *ro
 	stop := make(chan error)
 	stopped := uint32(0)
 
+	netConnection.SetOnBadMessageHandler(func(err error) {
+		atomic.AddUint32(&stopped, 1)
```
Comment (someone235) [inline protocol/protocol.go]: check the atomic before stopping

File: netadapter/server/grpcserver/grpc_server.go [hunk 3]
```diff
@@ -108,3 +110,15 @@ func (s *gRPCServer) Connect(address string) (server.Connection, error) {
...
+// IsBanned checks whether the given address had previously
+// been banned
+func (s *gRPCServer) IsBanned(address net.Addr) bool {
+	_, ok := s.bannedAddresses[address]
+	return ok
+}
+
+// Ban prevents the given address from connecting
+func (s *gRPCServer) Ban(address net.Addr) {
+	s.bannedAddresses[address] = struct{}{}
```
Comment (someone235) [inline netadapter/server/grpcserver/grpc_server.go:123]: Don't we want to strip out the port?

File: netadapter/server/grpcserver/p2pserver.go [hunk 4]
```diff
@@ -23,6 +23,12 @@ func (p *p2pServer) MessageStream(stream protowire.P2P_MessageStreamServer) erro
 	if !ok {
 		return errors.Errorf("Error getting stream peer info from context")
 	}
+
+	if p.server.IsBanned(peerInfo.Addr) {
```
Comment (someone235) [inline netadapter/server/grpcserver/p2pserver.go:27]: Use for outgoing connections as well

File: connmanager/connmanager.go (hunk 1 above)
Comment (stasatdaglabs) [reply connmanager/connmanager.go:118]: Done.

File: netadapter/server/grpcserver/grpc_server.go [hunk 5]
```diff
@@ -28,7 +28,7 @@ func NewGRPCServer(listeningAddrs []string) (server.Server, error) {
 	s := &gRPCServer{
 		server:          grpc.NewServer(),
 		listeningAddrs:  listeningAddrs,
-		bannedAddresses: make(map[net.Addr]struct{}),
+		bannedAddresses: make(map[*net.IP]struct{}),
```
Comment (someone235) [inline netadapter/server/grpcserver/grpc_server.go:31]: Don't we need it by value?

File: protocol/protocol.go [hunk 6]
```diff
@@ -28,8 +28,7 @@ func (m *Manager) routerInitializer(netConnection *netadapter.NetConnection) (*r
 		if err != nil {
 			if protocolErr := &(protocolerrors.ProtocolError{}); errors.As(err, &protocolErr) {
 				if protocolErr.ShouldBan {
-					// TODO(libp2p) Ban peer
-					panic("unimplemented")
+					m.context.ConnectionManager().Ban(netConnection)
```
Comment (svarogg) [inline protocol/protocol.go:31]: This is the only place where you call Ban.
And netAdapter is available here.
And ConnectionManager().Ban does NetAdapter.Ban()

So remove ConnectionManager.Ban and call NetAdapter.Ban() here?

File: netadapter/netconnection.go [hunk 7]
```diff
@@ -33,3 +33,9 @@ func (c *NetConnection) ID() *id.ID {
 func (c *NetConnection) Address() string {
 	return c.connection.Address().String()
 }
+
+// SetOnBadMessageHandler sets a handler function
+// for bad messages
+func (c *NetConnection) SetOnBadMessageHandler(onBadMessageHandler server.OnBadMessageHandler) {
```
Comment (svarogg) [inline netadapter/netconnection.go:39]: Messgae is bad... mkay?

But seriously, be a bit more specific as to what a "bad" message is? Is it malformed? Invalid? Unparseable?

File: protocol/protocol.go (hunk 6 above)
Comment (stasatdaglabs) [reply protocol/protocol.go:31]: We spoke about it and decided to leave it as is for now.

**Pull request merged**